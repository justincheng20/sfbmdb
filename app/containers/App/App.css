.title {
  text-align: center;
}

.button {
  text-decoration: none;
  font-size: 20px;
  color: blue;
}