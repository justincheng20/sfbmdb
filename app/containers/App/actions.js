import {
  LOAD_SANDWICHES,
  LOAD_SANDWICHES_SUCCESS,
  LOAD_SANDWICHES_ERROR,
  LOAD_DETAILS,
  LOAD_DETAILS_SUCCESS,
  LOAD_DETAILS_ERROR,
  LOAD_ALBUMS,
  LOAD_ALBUMS_SUCCESS,
  LOAD_ALBUMS_ERROR,
} from './constants';

export function loadSandwiches() {
  return {
    type: LOAD_SANDWICHES,
  };
}

export function sandwichesLoaded(sandwiches) {
  return {
    type: LOAD_SANDWICHES_SUCCESS,
    sandwiches,
  };
}

export function sandwichesLoadingError(error) {
  return {
    type: LOAD_SANDWICHES_ERROR,
    error,
  };
}

export function loadDetails(id) {
  return {
    type: LOAD_DETAILS,
    id,
  };
}

export function detailsLoaded(details, id) {
  return {
    type: LOAD_DETAILS_SUCCESS,
    details,
    id,
  };
}

export function detailsLoadingError(error) {
  return {
    type: LOAD_DETAILS_ERROR,
    error,
  };
}

export function loadAlbums(title) {
  return {
    type: LOAD_ALBUMS,
    title,
  };
}

export function albumsLoaded(albums) {
  return {
    type: LOAD_ALBUMS_SUCCESS,
    albums,
  };
}

export function albumsLoadingError(error) {
  return {
    type: LOAD_ALBUMS_ERROR,
    error,
  };
}
