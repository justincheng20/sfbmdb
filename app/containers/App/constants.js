export const LOAD_ITEMS = 'boilerplate/App/LOAD_ITEMS';
export const LOAD_ITEMS_SUCCESS = 'boilerplate/App/LOAD_ITEMS_SUCCESS';
export const LOAD_ITEMS_ERROR = 'boilerplate/App/LOAD_ITEMS_ERROR';
export const ADD_ITEM = 'boilerplate/App/ADD_ITEMS';
export const ADD_ITEM_SUCCESS = 'boilerplate/App/ADD_ITEMS_SUCCESS';
export const ADD_ITEM_ERROR = 'boilerplate/App/ADD_ITEMS_ERROR';
