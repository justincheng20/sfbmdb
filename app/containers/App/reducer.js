import produce from 'immer';
import {
  LOAD_ITEMS_SUCCESS,
  LOAD_ITEMS,
  LOAD_ITEMS_ERROR,
  ADD_ITEM,
  ADD_ITEM_SUCCESS,
  ADD_ITEM_ERROR,
} from './constants';

export const initialState = {
  loading: false,
  error: false,
  data: {
    items: false,
  },
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_ITEMS:
        draft.error = false;
        break;

      case LOAD_ITEMS_SUCCESS:
        draft.data = { items: action.items };
        draft.loading = false;
        break;

      case LOAD_ITEMS_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;

      case ADD_ITEM:
        draft.loading = true;
        draft.error = false;
        break;

      case ADD_ITEM_SUCCESS:
        if (state.data.items) {
          draft.data.items = [action.item, ...state.data.items];
        } else {
          draft.data.items = [action.item];
        }
        draft.loading = false;
        break;

      case ADD_ITEM_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;
    }
  });

export default appReducer;
