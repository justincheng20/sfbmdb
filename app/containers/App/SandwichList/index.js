import React, { useEffect, memo } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import { useInjectSaga } from 'utils/injectSaga';
import {
  makeSelectData,
  makeSelectLoading,
  makeSelectError,
} from 'containers/App/selectors';
import List from 'components/List';
import { Link } from 'react-router-dom';
import messages from './messages';
import { loadItems } from '../App/actions';
import saga from './saga';

const key = 'sandwiches';

export function SandwichList({ loading, error, onLoad, data }) {
  useInjectSaga({ key, saga });

  useEffect(() => {
    if (!data.items) onLoad();
  }, []);

  const listProps = {
    data,
    loading,
    error,
  };

  if (loading) {
    return 'Loading...';
  }

  if (error) {
    return 'Something went wrong.';
  }

  return (
    <div className="title">
      <h1>
        <FormattedMessage {...messages.header} />
      </h1>
      <Link className="button" to="/new">
        Add a new entry
      </Link>
      <List {...listProps} />
    </div>
  );
}

SandwichList.propTypes = {
  loading: PropTypes.bool,
  error: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
  data: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
  onLoad: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  data: makeSelectData(),
  loading: makeSelectLoading(),
  error: makeSelectError(),
});

export function mapDispatchToProps(dispatch) {
  return {
    onLoad: () => dispatch(loadItems()),
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
  memo,
)(SandwichList);
