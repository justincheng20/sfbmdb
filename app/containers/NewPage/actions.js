import { CHANGE_ITEM_NAME } from './constants';

export function changeItemName(itemName) {
  return {
    type: CHANGE_ITEM_NAME,
    itemName,
  };
}
