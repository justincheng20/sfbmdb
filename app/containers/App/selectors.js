import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const selectRouter = state => state.router;

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectItems = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.data.items,
  );

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

const makeSelectData = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.data,
  );

export {
  selectGlobal,
  makeSelectLoading,
  makeSelectError,
  makeSelectItems,
  makeSelectLocation,
  makeSelectData,
};
