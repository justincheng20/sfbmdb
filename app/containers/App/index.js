import React from 'react';
import { Switch, Route } from 'react-router-dom';

import HomePage from 'containers/HomePage/Loadable';
import NewPage from 'containers/NewPage/Loadable';
import SandwichPage from 'containers/SandwichPage/Loadable';
import SandwichList from 'containers/SandwichList/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';

import GlobalStyle from '../../global-styles';

export default function App() {
  return (
    <div>
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route exact path="/new" component={NewPage} />
        <Route exact path="/sandwiches" component={SandwichList} />
        <Route exact path="/sandwiches/:id" component={SandwichPage} />
        <Route component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </div>
  );
}
