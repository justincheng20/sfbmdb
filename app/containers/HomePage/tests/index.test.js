import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory, BrowserRouter } from 'react-router-dom';
import { HomePage, mapDispatchToProps } from '../index';
import { loadItems } from '../../App/actions';
import configureStore from '../../../configureStore';

describe('<HomePage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <BrowserRouter>
            <HomePage loading={false} error={false} data={{ items: [] }} />
          </BrowserRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should fetch items on mount if data is false', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <BrowserRouter>
            <HomePage data={{ items: false }} onLoad={submitSpy} />
          </BrowserRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(submitSpy).toHaveBeenCalled();
  });

  it('should not fetch items on mount if data is an array', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <BrowserRouter>
            <HomePage data={{ items: [] }} onLoad={submitSpy} />
          </BrowserRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(submitSpy).not.toHaveBeenCalled();
  });

  describe('mapDispatchToProps', () => {
    describe('onLoad', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onLoad).toBeDefined();
      });

      it('should dispatch loadItems when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        result.onLoad();
        expect(dispatch).toHaveBeenCalledWith(loadItems());
      });
    });
  });
});
