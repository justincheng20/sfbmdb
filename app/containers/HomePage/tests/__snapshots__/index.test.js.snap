// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<div
  class="title"
>
  <h1>
    <span>
      List of Items
    </span>
  </h1>
  <a
    class="button"
    href="/new"
  >
    Add a new item
  </a>
  <div>
    <ul
      class="StyledItemList-xvt1sj-0 fXsYHv"
    >
      <ul>
        <li
          class="StyledItem-sc-1crlcl1-0 khrTxW"
        >
          <div>
            "Nothing Added Yet"
          </div>
        </li>
      </ul>
    </ul>
  </div>
</div>
`;
